### PR TITLE
feat(ml+api): add FastAPI ML service, /api/reports endpoint, training pipeline & advice

### PR DESCRIPTION
**Pull request by ananthkt20**
PR Description (copy all of this)
Summary

This PR adds an end-to-end SMS smishing/spam classification flow:

A FastAPI (Uvicorn) microservice that loads a trained TF-IDF + Logistic Regression model and exposes /predict.

A new backend endpoint POST /api/reports that calls the ML service, normalizes/validates the result, stores a Report in MongoDB, and returns a user-friendly response with confidence and advice.

A training script to retrain the model from DatasetCombined.csv and export model.joblib + meta.json.

Small improvements to env configuration, model schema, and data normalization.

This enables the Android “Community → Report & Protect” screen to work against our backend (via BuildConfig.SERVERIP) and the local ML service on http://10.0.2.2:8000 (emulator) or http://localhost:8000 (server).

What changed
New files

machine-learning/ml_service.py – FastAPI app /predict (loads model.joblib, returns {label, confidence, probabilities, model_version}).

machine-learning/train.py – Training pipeline: reads datasets/DatasetCombined.csv, cleans labels (ham/spam/smishing), trains TF-IDF + LR, saves model + meta.

machine-learning/requirements.txt – Minimal deps for the ML service (fastapi, uvicorn, scikit-learn, pandas, joblib, etc.).

src/services/ml.service.js – Node client that calls the FastAPI /predict with timeouts/retries.

Modified files

src/index.js – Adds POST /api/reports endpoint:

Validates messageText (phone optional)

Calls ML service → normalizes → stores Report

Returns {classification, analysis, advice}

src/models/report.model.js – Adds classification object:

classification: {
  label: { type: String, enum: ['ham','spam','smishing'], required: true },
  confidence: { type: Number },   // 0–1
  badge: String,                   // Safe/Spam/Smishing
  severity: String,                // low/medium/high
  probabilities: {},               // optional
  model_version: String,
  advice: String,
  actions: [String],
}

(optional if present) src/routes/report.route.js, src/controllers/report.controller.js – Wires the new route.

Environment

.env.example – add:

ML_SERVICE_URL=http://localhost:8000

(Android emulator hits http://10.0.2.2:8000 from the device side, but server runs at localhost:8000.)

API Contract
POST /api/reports

Body

{
  "phoneNumber": "optional string",
  "messageText": "required string",
  "url": "optional string",
  "source": "android|web|test"
}

Response

{
  "status": "classified",
  "reportId": "...",
  "createdAt": "2025-09-18T12:34:56.000Z",
  "classification": {
    "label": "ham|spam|smishing",
    "confidence": 0.93,
    "badge": "Smishing",
    "severity": "high",
    "probabilities": { "...": 0.12 },
    "model_version": "v169..",
    "advice": "text...",
    "actions": ["...","..."]
  },
  "analysis": { "riskScore": 40, "tags": ["has_link"] }
}

FastAPI /predict (internal)

POST http://localhost:8000/predict

{ "text": "message to classify" }

Response

{ "label": "smishing", "confidence": 0.91, "probabilities": {...}, "model_version": "v..." }

How to run locally

ML service

cd machine-learning
python -m uvicorn ml_service:app --host 0.0.0.0 --port 8000 --reload

(Optional first-time) python train.py to generate model/model.joblib.

Node backend

.env has ML_SERVICE_URL=http://localhost:8000

npm i

npm run start (or your project start command)

Android

BuildConfig.SERVERIP = "http://10.0.2.2:3000" (already set)

Open app → Community → Report → test messages.

Notes for reviewers

Backward compatible: Only additive changes; existing routes untouched.

Normalization: The backend maps numeric/string labels and badges to the strict enum. If the ML model changes labels, it still maps to ham/spam/smishing.

Heuristics: Kept light, only for transparency in the response (analysis).

Security: Rate limiter + existing security middleware still apply on /api/reports.

Testing

curl sanity test:

curl -X POST http://localhost:3000/api/reports \
  -H "Content-Type: application/json" \
  -d '{"phoneNumber":"0457895623","messageText":"Your account is locked, click http://bad.link to verify."}'

Try a few:

Smishing: “Your bank account is locked. Verify now: http://bank-secure.example”

Spam: “Free trial — buy now! Limited time only.”

Safe: “Hey, are we still meeting at 6 pm?”

Risks & mitigations

Model drift: retrain via machine-learning/train.py when dataset updates.

Service down: backend falls back to lightweight heuristics (low confidence).

Env mismatch: documented ML_SERVICE_URL and Android’s 10.0.2.2.

### CODE REVIEW

**Comment by ananthkt20** [comment]
“Adds FastAPI ML microservice client, /api/reports endpoint with guardrails, training pipeline, and docs. Tested via Android app Community → Report.”

**Comment by ananthkt20** [comment]
## Summary
Adds FastAPI ML microservice client, `/api/reports` endpoint with guardrails, training pipeline, and advice text.
Tested via Android Community → Report screen.

## Key changes
- machine-learning/ml_service.py (FastAPI client)
- machine-learning/train.py, requirements.txt (training + deps)
- src/index.js (new /api/reports, normalization + guardrails)
- src/services/ml.service.js (HTTP client)
- src/models/report.model.js (classification schema)
- src/routes/report.route.js, src/controllers/report.controller.js

## How to test
1) Start ML: `python -m uvicorn ml_service:app --host 0.0.0.0 --port 8000 --reload`
2) Start Node server: `npm run dev` (PORT=3000)
3) POST to `POST /api/reports` with:
   `{ "messageText": "Your account is locked, verify at http://bad.link" }`
   Expect `{ classification: { label: "smishing" ... } }`.

## Notes
- Guardrails improve spam/smishing detection on promo/URL/OTP cases.
- No breaking changes to existing routes.

## Risks / Rollback
Minimal—new route + model fields. Rollback by reverting this PR.
